pkg/runner/action.go

```diff
@@ -381,63 +382,123 @@ func execAsComposite(step actionStep) common.Executor {
 				inputs[k] = eval.Interpolate(v)
 			}
 		}
-		// Doesn't work with the command processor has a pointer to the original rc
-		// compositerc := rc.Clone()
-		// Workaround start
-		backup := *rc
-		defer func() { *rc = backup }()
-		*rc = *rc.Clone()
-		scriptName := backup.CurrentStep
-		for rcs := &backup; rcs.Parent != nil; rcs = rcs.Parent {
-			scriptName = fmt.Sprintf("%s-composite-%s", rcs.Parent.CurrentStep, scriptName)
-		}
-		compositerc := rc
-		compositerc.Parent = &RunContext{
-			CurrentStep: scriptName,
+
+		env := make(map[string]string)
+		for k, v := range step.getStepModel().Environment() {
+			env[k] = eval.Interpolate(v)
 		}
-		// Workaround end
-		compositerc.Composite = action
-		envToEvaluate := mergeMaps(compositerc.Env, step.getStepModel().Environment())
-		compositerc.Env = make(map[string]string)
-		// origEnvMap: is used to pass env changes back to parent runcontext
-		origEnvMap := make(map[string]string)
-		for k, v := range envToEvaluate {
-			ev := eval.Interpolate(v)
-			origEnvMap[k] = ev
-			compositerc.Env[k] = ev
+
+		// run with the global config but without secrets
+		configCopy := *rc.Config
+		configCopy.Secrets = nil
+
+		// create a run context for the composite action to run in
+		compositerc := &RunContext{
+			Name:    rc.Name,
+			JobName: rc.JobName,
+			Run: &model.Run{
+				JobID: "composite-job",
+				Workflow: &model.Workflow{
+					Name: rc.Run.Workflow.Name,
+					Jobs: map[string]*model.Job{
+						"composite-job": {},
+					},
+				},
+			},
+			Config:           &configCopy,
+			StepResults:      map[string]*model.StepResult{},
+			JobContainer:     rc.JobContainer,
+			Inputs:           inputs,
+			ActionPath:       containerActionDir,
+			ActionRepository: rc.ActionRepository,
+			ActionRef:        rc.ActionRef,
+			Env:              env,
+			Masks:            rc.Masks,
```

Comment:
You forget the extrapath? My clone method would create a flat nearly 1:1 copy even after someone would add new properties.
```suggestion
			Masks:            rc.Masks,
			ExtraPath:        rc.ExtraPath,
```

Reply:
Yes, I missed that. I removed the clone method, because I felt a bit unsure on that. It did a shallow clone and keept most attributes of the struct. That means it was possible to manipulate references from the clone to the outer RC.
This could have been done very easy without notice. Having this more explicit might lead to miss an update here, but that is easier to track down.

Reply:
If you think it's better to clone, I could re-add it.